components/cam/src/dynamics/se/dp_coupling.F90

```diff
@@ -1,543 +1,623 @@
+    integer                  :: idmb2(1)                  ! column index within block
+    integer                  :: idmb3(1)                  ! local block id
+    integer                  :: tsize                     ! # vars per grid point passed to physics
+    integer,     allocatable :: bpter(:,:)                ! offsets into block buffer for packing 
+    integer                  :: cpter(pcols,0:pver)       ! offsets into chunk buffer for unpacking 
+    integer                  :: nphys, nphys_sq           ! physics grid parameters
+    real (kind=real_kind)    :: temperature(np,np,nlev)   ! Temperature from dynamics
+    ! Temp arrays for writing initial condition file when physics is on FV grid
+    real(kind=real_kind), allocatable :: ps_tmp_ic (:)
+    real(kind=real_kind), allocatable :: uv_tmp_ic (:,:,:)
+    real(kind=real_kind), allocatable :: T_tmp_ic  (:,:)
+    real(kind=real_kind), allocatable :: q_tmp_ic  (:,:,:)
     ! Frontogenesis
-    real (kind=real_kind), allocatable :: frontgf(:,:,:) ! temporary arrays to hold frontogenesis
-    real (kind=real_kind), allocatable :: frontga(:,:,:) !   function (frontgf) and angle (frontga)
-    ! Pointers to pbuf
-    real (kind=r8),        pointer     :: pbuf_frontgf(:,:)
-    real (kind=r8),        pointer     :: pbuf_frontga(:,:)
-
-    integer :: ncols,i,j,ierr
-
-    integer (kind=int_kind)  :: ioff, m
-    integer :: pgcols(pcols), idmb1(1), idmb2(1), idmb3(1)
-    integer :: tsize                 ! amount of data per grid point passed to physics
-    integer :: bpter(npsq,0:pver)    ! offsets into block buffer for packing data
-    integer :: cpter(pcols,0:pver)   ! offsets into chunk buffer for unpacking data
-
-    real (kind=real_kind), allocatable, dimension(:) :: bbuffer, cbuffer ! transpose buffers
-    integer :: tl_f
-
-    type(physics_buffer_desc), pointer :: pbuf_chnk(:)
-
-    real (kind=real_kind)    :: temperature(np,np,nlev) 
-    !----------------------------------------------------------------------
+    real (kind=real_kind), allocatable :: frontgf(:,:,:)  ! frontogenesis function
+    real (kind=real_kind), allocatable :: frontga(:,:,:)  ! frontogenesis angle 
+    real (kind=r8),            pointer :: pbuf_frontgf(:,:)
+    real (kind=r8),            pointer :: pbuf_frontga(:,:)
+    ! Transpose buffers
+    real (kind=real_kind), allocatable, dimension(:) :: bbuffer 
+    real (kind=real_kind), allocatable, dimension(:) :: cbuffer
+    !---------------------------------------------------------------------------
 
     nullify(pbuf_chnk)
     nullify(pbuf_frontgf)
     nullify(pbuf_frontga)
 
-    if (use_gw_front) then
+    if (fv_nphys > 0) then
+      nphys = fv_nphys
+    else
+      nphys = np
+    end if
+    nphys_sq = nphys*nphys
 
-       allocate(frontgf(npsq,pver,nelemd), stat=ierr)
-       if (ierr /= 0) call endrun("dp_coupling: Allocate of frontgf failed.")
+    ! Allocate temporary arrays to hold data for physics decomposition
```

Comment:
It seems that memory will be allocated each coupling step. Could allocation be moved to some init? Is it actually better to allocate and free each time tor to call allocate once?

Reply:
I'm not sure what's the better approach. We do this type of thing pretty often in SP-E3SM, so I figured it wouldn't add much overhead.

Reply:
It's true that a small number of allocations once per physics step will not matter much compared with the cost of the physics step itself. Allocations in a low-level routine called many times in a step would of course be much worse.

But it may be easy to keep stack allocation here, so it might be worth doing that. Here's a full program demonstrating how one can do that:
```F90
module foo
  integer :: &
       sz1, sz2, & ! model of np and nphys
       ! max_sz is the size in the sz dimension needed to accommodate sz1 or sz2
       max_sz
contains
  subroutine init(n1,n2)
    ! Initialize this module by setting up sizes.
    integer, intent(in) :: n1,n2
    sz1 = n1
    sz2 = n2
    max_sz = max(sz1,sz2)
  end subroutine init

  subroutine bar(use_sz1)
    ! A function in this module can stack allocate arrays using max_sz.
    logical, intent(in) :: use_sz1
    integer :: arr(max_sz,12,3)  ! stack-allocate using max_sz

    ! Some pointless code just to exercise size vs capacity.
    if (use_sz1) then
       arr(1:sz1,:,:) = 1
    else
       arr(1:sz2,:,:) = 2
    end if
  end subroutine bar
end module foo

program main
  use foo, only: init, bar

  call init(3,2)
  call bar(.true.)
  call bar(.false.)
end program main
```
In your code, you would have something like `max_nphys_sq = max(np,nphys)**2` and then stack allocate with, e.g., `ps_tmp(max_npyhs_sq,nelemd)`.

Reply:
@ambrad, Thanks for the suggestion. I implemented it and it works just fine.